### PR TITLE
FKM algorithm for necklaces

### PR DESCRIPTION
**Pull request by Pontus-vB**
#### Brief description of what is fixed or changed
Faster generation of necklaces and bracelets by the FKM algorithm. See [Generating necklaces](https://doi.org/10.1016/0196-6774(92)90047-G) by Ruskey, Savage, and Wang.

#### Other comments
The current implementation of ```necklaces(n,k)``` is very inefficient; it goes through all $k^n$ $k$-ary sequences of length $n$, finds the minimal rotation of each, and filters out duplicates.

Timings (on an old laptop):
```from sympy.utilities.iterables import necklaces
from timeit import timeit
timeit('sum(1 for n in range(12) for k in range(12-n) for p in necklaces(n,k))',globals={'necklaces':necklaces},number=1)
# 21.757525700028054
timeit('sum(1 for n in range(12) for k in range(12-n) for p in necklaces(n,k))',globals={'necklaces':necklaces_FKM},number=1)
# 0.009573900024406612
timeit('sum(1 for n in range(12) for k in range(12-n) for p in necklaces(n,k,True))',globals={'necklaces':necklaces},number=1)
# 45.918704900017474
timeit('sum(1 for n in range(12) for k in range(12-n) for p in necklaces(n,k,True))',globals={'necklaces':necklaces_FKM},number=1)
# 0.03058190003503114
```
In these tests, the proposed version is named ```necklaces_FKM``` and the current version is ```necklaces```.

#### Release Notes

* utilities
  * Faster generation of necklaces and bracelets.

### CODE REVIEW

**Comment by Pontus-vB** [comment]
Not sure about the .mailmap file. My next guess would be "Name \<real mail adress\> \<no-reply mail adress\>" (without the user name). Or should the no-reply adress not be there at all? (Sorry about the confusion; this is my very first PR.)

**Comment by oscarbenjamin** [comment]
You can see the raw commit here:
https://github.com/sympy/sympy/commit/4074b35692da23c28712cdc35c323e2a231ef35b.patch
It has:
```
From: =?UTF-8?q?Pontus=20von=20Br=C3=B6mssen?=
 <118356995+Pontus-vB@users.noreply.github.com>
```
So maybe the .mailmap should be:
```
Pontus von Brömssen <pontus.vonbromssen+github@gmail.com> Pontus von Brömssen <118356995+Pontus-vB@users.noreply.github.com>
```
Or perhaps it should literally have the UTF-8 escapes like:
```
Pontus von Brömssen <pontus.vonbromssen+github@gmail.com> =?UTF-8?q?Pontus=20von=20Br=C3=B6mssen?= <118356995+Pontus-vB@users.noreply.github.com>
```
The point is to link the metadata in the commit with what you want to have in the AUTHORS file like:
```
Real Name <real email address> Git Name <git email address>
```
Ideally you would just set your name and email address with `git config` to be whatever they are supposed to be and then you only need to add those to the .mailmap:
```
Pontus von Brömssen <pontus.vonbromssen+github@gmail.com>
```
You can test whether this is working locally by running `bin/mailmap_check.py`.

**Comment by oscarbenjamin** [comment]
I don't know this function at all but it seems @smichr is the author and might review this.

All I can say is that it would be good to add a reference for the algorithm to the docstring. See
https://docs.sympy.org/latest/contributing/docstring.html#references-section

File: sympy/utilities/iterables.py [hunk 1]
```diff
@@ -2546,8 +2546,25 @@ def necklaces(n, k, free=False):
...
+    while 1:
+        i = n - 1
+        while a[i] == k - 1:
+            i -= 1
+            if i == -1:
+                return
+        a[i] += 1
+        for j in range(n - i - 1):
+            a[j + i + 1] = a[j]
+        if n % (i + 1) == 0 and (not free or all(a <= a[j::-1] + a[-1:j:-1] for j in range(n - 1))):
```
Comment (smichr) [inline sympy/utilities/iterables.py:2565]: I tested, without failure
```python
for n in range(6):
    for k in range(n+1):
        for b in range(2):
            assert all(i==j for i,j in zip(necklaces(n,k,b), old_necklaces(n,k,b)))
```
This is very nice!

I wonder if that `all(a <= ...` could be improved, perhaps, by finiding the minimal reversed string and comparing (using a better [`minlex'](#https://en.wikipedia.org/wiki/Lexicographically_minimal_string_rotation#Booth.27s_Algorithm))). But my testing only yields about a 10% improvement only at larger necklaces. The current code is straightforward and a nice improvement.

Comment (smichr) [reply sympy/utilities/iterables.py:2565]: +1 to adding reference

Comment (Pontus-vB) [reply sympy/utilities/iterables.py:2565]: Thanks. I’m not sure how to improve the bracelet test, so I’ll leave it as is. Will add the reference.

File: sympy/utilities/iterables.py [hunk 2]
```diff
@@ -2545,9 +2545,30 @@ def necklaces(n, k, free=False):
...
-    return uniq(minlex(i, directed=not free) for i in
-        variations(list(range(k)), n, repetition=True))
+    # The FKM algorithm
+    if k == 0 and n > 0:
+        return
+    a = [0]*n
+    yield tuple(a)
+    if n == 0:
+        return
+    while 1:
```
Comment (sylee957) [inline sympy/utilities/iterables.py:2560]:
```suggestion
    while True:
```

File: sympy/utilities/iterables.py [hunk 3]
```diff
@@ -2545,9 +2545,30 @@ def necklaces(n, k, free=False):
 
     .. [1] http://mathworld.wolfram.com/Necklace.html
 
+    .. [2] Frank Ruskey, Carla Savage, and Terry Min Yih Wang,
+        Generating necklaces, Journal of Algorithms 13 (1992), 414–430;
+        https://doi.org/10.1016/0196-6774(92)90047-G
```
Comment (smichr) [inline sympy/utilities/iterables.py:2550]:
```suggestion
    .. [2] Frank Ruskey, Carla Savage, and Terry Min Yih Wang,
        Generating necklaces, Journal of Algorithms 13 (1992), 414-430;
        https://doi.org/10.1016/0196-6774(92)90047-G
```

**Pull request merged**